### PR TITLE
Feature/stein thinning

### PR DESCRIPTION
**Pull request by tl82649**
### PR Type

- Feature
- Tests

### Description

Regularised Stein thinning algorithm added.

### How Has This Been Tested?

Unit, integration and performance tests.

### Checklist before requesting a review
- [x] I have made sure that my PR is not a duplicate.
- [x] My code follows the style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have performed a self-review of my code.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

File: .cspell/people.txt [hunk 1]
```diff
@@ -20,6 +21,9 @@ Rudi
 Sahaj
 Schreuder
 Smirnov
+Staber
+tein
```
Comment (tc85324) [inline .cspell/people.txt:25]: We can delete this by updating the bibliography slightly (see later comment)

File: documentation/source/references.bib [hunk 2]
```diff
@@ -60,3 +60,11 @@ @misc{chen2023randomly
   archiveprefix = {arXiv},
   primaryclass  = {math.NA}
 }
+
+@inproceedings{benard2023kernel,
+  title     = {Kernel {S}tein Discrepancy thinning: a theoretical perspective of pathologies and a practical fix with regularization},
```
Comment (tc85324) [inline documentation/source/references.bib:65]: Although bibtex does suggest the style `{S}tein` for capitalization, to avoid confusing the spell checker, we can instead do the below with no loss of fuctionality (we still preserve the capitalization).
```suggestion
  title     = {Kernel {Stein} Discrepancy thinning: a theoretical perspective of pathologies and a practical fix with regularization},
```

File: coreax/coresubset.py [hunk 3]
```diff
@@ -151,6 +153,9 @@ def fit_to_size(self, coreset_size: int) -> None:
 
         :param coreset_size: The size of the of coreset to generate
         """
+        if coreset_size == 0:
+            raise ValueError("coreset_size must be non-zero")
+
```
Comment (tc85324) [inline coreax/coresubset.py:158]: Can this be removed if we make the suggested changes to line 187 (see later comment)

File: coreax/coresubset.py [hunk 4]
```diff
@@ -617,3 +620,294 @@ def _loop_body(
...
+        We first compute a score function, and then the Stein kernel. This is used to
+        greedily choose points in the coreset to minimise kernel Stein discrepancy
+        (KSD).
+
+        :param coreset_size: The size of the of coreset to generate
+        """
+        if coreset_size == 0:
+            raise ValueError("coreset_size must be non-zero")
+
```
Comment (tc85324) [inline coreax/coresubset.py:737]: Same comment as in https://github.com/gchq/coreax/pull/572#discussion_r1579007131

File: coreax/coresubset.py [hunk 5]
```diff
@@ -617,3 +620,294 @@ def _loop_body(
...
+                raise ValueError("coreset_size must be non-zero") from exception
+            raise
+
+        # Assign coreset indices & coreset to original data object
+        self.coreset_indices = coreset_indices
+        self.coreset = self.original_data.pre_coreset_array[self.coreset_indices, :]
+
+    @staticmethod
+    @partial(jit, static_argnames=["kernel_vectorised", "unique"])
+    def _greedy_body(
```
Comment (tc85324) [inline coreax/coresubset.py:832]: This function repeats a lot of what is in the `_greedy_body` of `KernelHerding`. We can reduce the duplication by having `_greedy_body(..., selection_function: Callable[[Array], Array])`, where  `index_to_include_in_coreset = selection_function(current_kernel_similarity_penalty)`. This would require `_greedy_body` to become a module level function.

File: tests/unit/test_coresubset.py [hunk 6]
```diff
@@ -1426,6 +1429,361 @@ def test_rp_cholesky_fit_invalid_data(self):
...
+
+        # Apply SteinThinning on the dataset, and record the coreset for comparison
+        coresubset_object_not_random.fit(
+            original_data=data, strategy=coreax.reduction.SizeReduce(self.coreset_size)
+        )
+        fitted_coresubset = coresubset_object_not_random.coreset
+
+        # Create a random refinement object and generate a coreset, for comparison
+        coresubset_object_random = coreax.coresubset.RandomSample(
+            random_key=self.random_key, unique=True
```
Comment (tc85324) [inline tests/unit/test_coresubset.py:1594]: Should random be `unique=True` in this test, or should it match `SteinThining` and have `unique=False`?
```suggestion
            random_key=self.random_key, unique=False
```

File: tests/unit/test_coresubset.py [hunk 7]
```diff
@@ -1426,6 +1429,361 @@ def test_rp_cholesky_fit_invalid_data(self):
...
+        """
+        Test the non-unique fit method of the SteinThinning class with a simple example.
+
+        The test checks that a coreset generated via SteinThinning has an improved
+        quality (measured by maximum mean discrepancy) than one generated by random
+        sampling. We further check that if the coreset generated by SteinThinning is
+        refined, the quality improves yet again.
+        """
+        # Define specific test instance setup
+        # kernel = coreax.kernel.SquaredExponentialKernel()
```
Comment (tc85324) [inline tests/unit/test_coresubset.py:1564]: Commented out code needs removal.
```suggestion
```

File: tests/unit/test_coresubset.py [hunk 8]
```diff
@@ -1426,6 +1429,361 @@ def test_rp_cholesky_fit_invalid_data(self):
...
+    def test_fit_comparison_to_random_and_refined(self) -> None:
+        """
+        Test the fit method of the SteinThinning class with a simple example.
+
+        The test checks that a coreset generated via SteinThinning has an improved
+        quality (measured by maximum mean discrepancy) than one generated by random
+        sampling. We further check that if the coreset generated by SteinThinning is
+        refined, the quality improves yet again.
+        """
+        # kernel = coreax.kernel.SquaredExponentialKernel()
```
Comment (tc85324) [inline tests/unit/test_coresubset.py:1504]: Commented out code needs removal.
```suggestion
```

**Review by tc85324 — COMMENTED** [review]
I've now looked at the tests and done a first pass on mathematical correctness. I've got a few questions which might require opening new issues to be resolved in separate pulls, but really nothing major to change here.

File: .cspell/people.txt (hunk 1 above)
Comment (tl82649) [reply .cspell/people.txt]: Done.

File: documentation/source/references.bib (hunk 2 above)
Comment (tl82649) [reply documentation/source/references.bib]: Done.

File: coreax/coresubset.py (hunk 3 above)
Comment (tl82649) [reply coreax/coresubset.py]: Done for herding. However, Stein thinning does need a check-for-nonpositive-integer if regularisation is used (see below).

File: coreax/coresubset.py (hunk 4 above)
Comment (tl82649) [reply coreax/coresubset.py]: I've kept this one as the regularised version of Stein thinning needs to check for division by zero when setting the $\lambda$ parameter. The exception doesn't fire as is due to `jnp.zeros` accepting `0` as an argument in the `try; catch`.

File: coreax/coresubset.py (hunk 5 above)
Comment (tl82649) [reply coreax/coresubset.py:829]: I agree. I haven't implemented it here as I think this would need to be a separate ticket.

File: coreax/coresubset.py [hunk 9]
```diff
@@ -617,3 +620,294 @@ def _loop_body(
...
+            score_function = coreax.score_matching.KernelDensityMatching(
+                self.kernel.length_scale, self.original_data.pre_coreset_array
+            ).match(None)
+        else:
+            score_function = self.score_method.match(
+                self.original_data.pre_coreset_array
+            )
+
+        # Create a Stein kernel
+        self.stein_kernel = coreax.kernel.SteinKernel(self.kernel, score_function)
```
[the remark this answers is not in the document]
Comment (tl82649) [reply coreax/coresubset.py]: Done.

File: coreax/coresubset.py [hunk 10]
```diff
@@ -617,3 +620,294 @@ def _loop_body(
...
+    compressed representation.
+
+    Given one has selected :math:`T` data points for their compressed representation of
+    the original dataset, (regularised) Stein thinning selects the next point as:
+
+    .. math::
+
+        x_{T+1} = \arg\min_{x} \left( k_P(x, x) / 2 + \Delta^+ \log p(x) -
+            \lambda T \log p(x) + \frac{1}{T+1}\sum_{t=1}^T k_P(x, x_t) \right)
+
```
[the remark this answers is not in the document]
Comment (tl82649) [reply coreax/coresubset.py:635]: Yes, correct, it should. Done. Also updated the actual code to reflect this.

File: tests/unit/test_coresubset.py [hunk 11]
```diff
@@ -391,7 +392,9 @@ def mock_kernel_vectorised(_, __):
             # Since we have unique set to True in the greedy body call, we should have
             # set the penalty for point index 1 to be infinite
             np.testing.assert_array_equal(
-                kernel_similarity_penalty_1, np.asarray([0.5, np.inf, 0.5])
+                # kernel_similarity_penalty_1, np.asarray([0.5, np.inf, 0.5])
```
[the remark this answers is not in the document]
Comment (tl82649) [reply tests/unit/test_coresubset.py]: Done.

File: tests/unit/test_coresubset.py (hunk 6 above)
Comment (tl82649) [reply tests/unit/test_coresubset.py]: Done.

File: tests/unit/test_coresubset.py (hunk 7 above)
Comment (tl82649) [reply tests/unit/test_coresubset.py]: Done.

File: tests/unit/test_coresubset.py [hunk 12]
```diff
@@ -1426,6 +1429,361 @@ def test_rp_cholesky_fit_invalid_data(self):
...
+        # Define some data - sufficiently large that we would not expect a random sample
+        # to typically compete with a SteinThinning approach
+        generator = np.random.default_rng(self.random_data_generation_key)
+        x = generator.random((num_data_points, self.dimension))
+        data = coreax.data.ArrayData.load(x)
+        num_samples_length_scale = min(num_data_points, 1_000)
+        idx = generator.choice(num_data_points, num_samples_length_scale, replace=False)
+        length_scale = float(coreax.kernel.median_heuristic(x[idx]))
+        # Define specific test instance setup
+        # kernel = coreax.kernel.SquaredExponentialKernel(length_scale=length_scale)
```
[the remark this answers is not in the document]
Comment (tl82649) [reply tests/unit/test_coresubset.py]: Done.

File: tests/unit/test_coresubset.py (hunk 8 above)
Comment (tl82649) [reply tests/unit/test_coresubset.py]: Done.

File: tests/integration/test_herding_basic_weighted.py [hunk 13]
```diff
@@ -61,6 +63,13 @@ def test_herding_basic(self) -> None:
                     herding coreset MMD",
             )
 
+            self.assertLess(
+                mmd_stein,
+                mmd_random,
+                msg="MMD for random sampling was unexpectedly lower than \
+                    Stein thinning coreset MMD",
+            )
```
[the remark this answers is not in the document]
Comment (tl82649) [reply tests/integration/test_herding_basic_weighted.py:71]: Yes. I've raised a separate ticket to remove the comparison with random from `tests/unit`, since there's a chance that random is better depending on data generation and PNRG settings. There's an argument that they should remain in `tests/integration`, but we haven't decided on a better approach to the above yet.

File: coreax/coresubset.py [hunk 14]
```diff
@@ -617,3 +615,296 @@ def _loop_body(
...
+            # Note that a TypeError is raised if the size input to jnp.zeros is negative
+            coreset_indices = jnp.zeros(coreset_size, dtype=jnp.int32)
+
+        except TypeError as exception:
+            if coreset_size <= 0 or isinstance(coreset_size, float):
+                raise ValueError(
+                    "coreset_size must be a positive integer"
+                ) from exception
+            raise
+
```
Comment (tc85324) [inline coreax/coresubset.py:767]: I think this try-except is now redundant due to the check on line 731

File: coreax/coresubset.py [hunk 15]
```diff
@@ -617,3 +615,296 @@ def _loop_body(
...
+            coreset_indices, kernel_similarity_penalty = lax.fori_loop(
+                lower=0,
+                upper=coreset_size,
+                body_fun=body,
+                init_val=(coreset_indices, kernel_similarity_penalty),
+            )
+        except IndexError as exception:
+            if coreset_size == 0:
+                raise ValueError("coreset_size must be non-zero") from exception
+            raise
```
Comment (tc85324) [inline coreax/coresubset.py:821]: I think this try-except is also redundant due to the check on line 731

File: coreax/coresubset.py (hunk 5 above)
Comment (tc85324) [reply coreax/coresubset.py:829]: We can lump this into the #581 refactor

File: coreax/coresubset.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (tc85324) [reply coreax/coresubset.py]: Just to make sure this isn't lost in any future changes, please can you use a `SteinKernel` in one of the test cases.

**Review by tc85324 — CHANGES_REQUESTED** [review]
Some very minor changes to remove (what I think is) dead code, but otherwise this looks good to me!

File: coreax/coresubset.py (hunk 14 above)
Comment (tl82649) [reply coreax/coresubset.py:759]: Done.

File: coreax/coresubset.py (hunk 15 above)
Comment (tl82649) [reply coreax/coresubset.py]: Done.

File: coreax/coresubset.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (tl82649) [reply coreax/coresubset.py]: I've added a test that checks that the Stein kernel is being set correctly.

**Review by tc85324 — APPROVED** [review]
Looks good to me!
